File: Hackerrank/3_month_preparation/python/week3.py
def permuting_two_arrays(k, A, B):
    # Time complexity: O(n)
    # Space complexity (ignoring input): O(n)
    A = sorted(A)
    B = sorted(B, reverse=True)

    for index in range(0, len(A)):
        if (A[index] + B[index]) < k:
            return "NO"

    return "YES"


def subarray_division_2(s, d, m):
    # Time complexity: O(n), you could speed up creating a sum variable, but still O(n)
    # Space complexity (ignoring input): O(1)
    ways_to_divide = 0
    for index in range(0, len(s) - m + 1):
        subarray = s[index : index + m]
        if sum(subarray) == d:
            ways_to_divide += 1

    return ways_to_divide


def strings_xor(s, t):
    res = ""
    for i in range(len(s)):
        if s[i] == t[i]:
            res += "0"
        else:
            res += "1"

    return res


def sales_by_match(n, ar):
    # Time complexity: O(n)
    # Space complexity (ignoring input): O(n)
    socks_dict = {}
    for sock in ar:
        if sock in socks_dict:
            socks_dict[sock] += 1
        else:
            socks_dict[sock] = 1

    total_pairs = 0
    for values in socks_dict.values():
        total_pairs += values // 2

    return total_pairs


def migratory_birds(arr):
    # Time complexity: O(n)
    # Space complexity (ignoring input): O(1). You could use a frequency array instead
    # of a dictionary and use even less space
    birds_dict = {}
    for bird in arr:
        if bird in birds_dict:
            birds_dict[bird] += 1
        else:
            birds_dict[bird] = 1

    most_frequent = [6, 0]
    for bird, frequency in birds_dict.items():
        if frequency > most_frequent[1]:
            most_frequent = [bird, frequency]
        if (frequency == most_frequent[1]) and (bird < most_frequent[0]):
            most_frequent[0] = bird

    return most_frequent[0]


def maximum_perimeter_triangle(sticks: list):
    # Time complexity: O(n*log(n))
    # Space complexity (ignoring input): O(1)
    sticks.sort(reverse=True)
    for index in range(0, len(sticks) - 2):
        if sticks[index] < sticks[index + 1] + sticks[index + 2]:
            triangle = [sticks[index], sticks[index + 1], sticks[index + 2]]
            triangle.sort()
            return triangle

    return [-1]


def zig_zag_sequence(a, n):
    a.sort()
    mid = int((n - 1) / 2)
    a[mid], a[n - 1] = a[n - 1], a[mid]

    st = mid + 1
    ed = n - 2
    while st <= ed:
        a[st], a[ed] = a[ed], a[st]
        st = st + 1
        ed = ed - 1

    for i in range(n):
        if i == n - 1:
            print(a[i])
        else:
            print(a[i], end=" ")
    return


def drawing_book(n, p):
    # Time complexity: O(1)
    # Space complexity (ignoring input): O(1)
    flips_front = p // 2
    flips_end = n // 2 - p // 2

    if flips_front > flips_end:
        return flips_end

    return flips_front

File: Hackerrank/3_month_preparation/rust/src/week3.rs
pub fn permuting_two_arrays(k: i32, A: &[i32], B: &[i32]) -> String {
    //Time complexity: O(n*log(n))
    //Space complexity (ignoring input): O(n)
    let mut A = A.to_vec();
    A.sort_unstable();
    let mut B = B.to_vec();
    B.sort_unstable_by(|x, y| y.cmp(x));
    for (index, a_value) in A.iter().enumerate() {
        let b_value = B[index];
        if a_value + b_value < k {
            return "NO".to_string();
        }
    }

    "YES".to_string()
}

pub fn subarray_division_2(s: &[i32], d: i32, m: i32) -> i32 {
    //Time complexity: O(n)
    //Space complexity (ignoring input): O(1)
    let mut ways_to_divide = 0;
    let mut sum: i32 = s[0..m as usize].iter().sum();
    if sum == d {
        ways_to_divide += 1
    }
    for index in 1..(s.len() - m as usize + 1) {
        sum += s[index + m as usize - 1] - s[index - 1];
        if sum == d {
            ways_to_divide += 1
        }
    }

    ways_to_divide
}

//XOR Strings 3
fn main() {
    read_input();
}

fn read_input() {
    let mut buffer = String::new();
    loop {
        let read_line_result = std::io::stdin().read_line(&mut buffer);
        match read_line_result {
            Ok(0) => break,
            Ok(_n) => (),
            Err(error) => println!("error: {:?}", error),
        }
    }
    treat_input(buffer)
}
fn treat_input(buffer: String) {
    let buffer = buffer.split('\n').collect::<Vec<_>>();
    xor_strings_3(buffer[0], buffer[1]);
}

fn xor_strings_3(s1: &str, s2: &str) {
    let _ = s1
        .chars()
        .zip(s2.chars())
        .map(|(a, b)| if a == b { print!("0") } else { print!("1") })
        .collect::<Vec<_>>();
}

fn sales_by_match(n: i32, ar: &[i32]) -> i32 {
    //Time complexity: O(n)
    //Space complexity (ignoring input): O(n)
    let mut socks_hash = std::collections::HashMap::new();
    for sock in ar {
        match socks_hash.get(sock) {
            Some(v) => socks_hash.insert(sock, v + 1),
            None => socks_hash.insert(sock, 1),
        };
    }
    let mut total_pairs = 0;
    for values in socks_hash.values() {
        total_pairs += values / 2;
    }
    total_pairs
}

fn migratory_birds(arr: &[i32]) -> i32 {
    //Time complexity: O(n)
    //Space complexity (ignoring input): O(1). You could use a frequency array instead
    //of a dictionary and use even less space
    let mut birds_hash = std::collections::HashMap::new();
    for &bird in arr {
        match birds_hash.get(&bird) {
            Some(value) => birds_hash.insert(bird, value + 1),
            None => birds_hash.insert(bird, 1),
        };
    }

    struct BirdInfo {
        bird: i32,
        frequency: i32,
    }
    let mut most_frequent = BirdInfo {
        bird: 6,
        frequency: 0,
    };
    for (bird, frequency) in birds_hash.iter() {
        if frequency > &most_frequent.frequency {
            most_frequent.bird = *bird;
            most_frequent.frequency = *frequency
        }
        if (frequency == &most_frequent.frequency) & (bird < &most_frequent.bird) {
            most_frequent.bird = *bird;
        }
    }
    most_frequent.bird
}

fn maximum_perimeter_triangle(sticks: &[i32]) -> Vec<i32> {
    //Time complexity: O(n*log(n))
    //Space complexity (ignoring input): O(n)
    let mut sorted_sticks = sticks.to_vec();
    sorted_sticks.sort_unstable_by(|a, b| b.cmp(a));
    for index in 0..(sorted_sticks.len() - 2) {
        if sorted_sticks[index] < sorted_sticks[index + 1] + sorted_sticks[index + 2] {
            let mut triangle = vec![
                sorted_sticks[index],
                sorted_sticks[index + 1],
                sorted_sticks[index + 2],
            ];
            triangle.sort_unstable();

            return triangle;
        }
    }

    vec![-1]
}

fn drawing_book(n: i32, p: i32) -> i32 {
    //Time complexity: O(1)
    //Space complexity (ignoring input): O(1)
    let flips_front = p / 2;
    let flips_end = n / 2 - p / 2;

    if flips_front > flips_end {
        return flips_end;
    }

    flips_front
}
